### ISSUE TITLE
v1.9.1 update failed: ncc: command not found

### ISSUE THREAD

**Issue by nero120** [nextcloud/nextcloudpi#827]
Tried to update via `sudo ncp-config`, this is the output I got:

```
Downloading updates
Performing updates
Running letsencrypt
Creating virtual environment...
Installing Python packages...
Installation succeeded.
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Plugins selected: Authenticator webroot, Installer None
Cert not yet due for renewal
Keeping the existing certificate

-------------------------------------------------------------------------------
Certificate not yet due for renewal; no action taken.
-------------------------------------------------------------------------------
/usr/local/bin/ncp/NETWORKING/letsencrypt.sh: line 93: ncc: command not found
```

Line 93 of that script refers to these commands:

```
# Configure Nextcloud
ncc config:system:set trusted_domains 4 --value=$DOMAIN
ncc config:system:set overwrite.cli.url --value=https://"$DOMAIN"/
```

I've never seen this `ncc` command, if I want to access the nextcloud cli I always use `sudo -u www-data php /var/www/nextcloud/occ` as per the [manual](https://docs.nextcloud.com/server/stable/admin_manual/configuration_server/occ_command.html?highlight=occ). What is `ncc`?

**System information**

<details>
<summary>NextCloudPi diagnostics</summary>

```

grep: /etc/mysql/mariadb.conf.d/90-ncp.cnf: No such file or directory
NextCloudPi version  v1.8.0
NextCloudPi image    NextCloudPi_09-12-17
distribution         Raspbian GNU/Linux 9 \n \l
automount            no
USB devices          sda sdb
datadir              /media/data/ncdata
data in SD           no
data filesystem      btrfs
data disk usage      50G/932G
rootfs usage         2.0G/28G
swapfile             /dev/mmcblk0p3
dbdir                 (doesn't exist)
Nextcloud check      ok
Nextcloud version    15.0.5.3
HTTPD service        up
PHP service          up
MariaDB service      up
Redis service        up
Postfix service      up
internet check       ok
port check 80        open
port check 443       open
IP                   ***REMOVED SENSITIVE VALUE***
gateway              ***REMOVED SENSITIVE VALUE***
interface            eth0
certificates         ***REMOVED SENSITIVE VALUE***
NAT loopback         no
uptime               51min

```
</details>

<details>
<summary>Nextcloud configuration</summary>

```

{
    "system": {
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost",
            "192.168.1.5",
            "nero120.cloud"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "https:\/\/nero120.cloud\/",
        "dbtype": "mysql",
        "version": "15.0.5.3",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "memcache.local": "\\OC\\Memcache\\Redis",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "timeout": 0,
            "password": "***REMOVED SENSITIVE VALUE***"
        },
        "mail_smtpmode": "smtp",
        "mail_smtpauthtype": "PLAIN",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "overwriteprotocol": "https",
        "loglevel": 0,
        "log_type": "file",
        "maintenance": false,
        "mail_smtpauth": 1,
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "587",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpsecure": "tls",
        "theme": "",
        "updater.release.channel": "stable",
        "logfile": "\/media\/data\/ncdata\/nextcloud.log",
        "htaccess.RewriteBase": "\/",
        "updater.secret": "***REMOVED SENSITIVE VALUE***"
    }
}

```
</details>

<details>
<summary>HTTPd logs</summary>

```

[Wed Mar 06 22:55:14.071629 2019] [http2:warn] [pid 1580] AH10034: The mpm module (prefork.c) is not supported by mod_http2. The mpm determines how things are processed in your server. HTTP/2 has more demands in this regard and the currently selected mpm will just not do. This is an advisory warning. Your server will continue to work, but the HTTP/2 protocol will be inactive.
[Wed Mar 06 22:55:15.022803 2019] [mpm_prefork:notice] [pid 1580] AH00163: Apache/2.4.25 (Raspbian) OpenSSL/1.0.2r configured -- resuming normal operations
[Wed Mar 06 22:55:15.022953 2019] [core:notice] [pid 1580] AH00094: Command line: '/usr/sbin/apache2'
No site matches ncp-activation
[Wed Mar 06 22:56:37.000273 2019] [php7:notice] [pid 1599] [client 192.168.1.3:55248] PHP Notice:  Undefined index: app in /var/www/ncp-web/index.php on line 219
grep: /etc/mysql/mariadb.conf.d/90-ncp.cnf: No such file or directory
grep: /etc/mysql/mariadb.conf.d/90-ncp.cnf: No such file or directory
find: '/etc/letsencrypt/live/': Permission denied
Wed Mar  6 23:09:31 2019 (3547): Error Cannot kill process 1599: Success!
[Wed Mar 06 23:09:32.042143 2019] [core:notice] [pid 1580] AH00052: child pid 3547 exit signal Segmentation fault (11)
No site matches ncp-activation
[Wed Mar 06 23:39:19.264317 2019] [php7:warn] [pid 3810] [client 192.168.1.3:59797] PHP Warning:  file_get_contents(/usr/local/etc/ncp-config.d/.letsencrypt.sh.cfg): failed to open stream: No such file or directory in /var/www/ncp-web/elements.php on line 183
[Wed Mar 06 23:39:19.264460 2019] [php7:warn] [pid 3810] [client 192.168.1.3:59797] PHP Warning:  sizeof(): Parameter must be an array or an object that implements Countable in /var/www/ncp-web/elements.php on line 190
[Wed Mar 06 23:39:19.287042 2019] [php7:warn] [pid 3810] [client 192.168.1.3:59797] PHP Warning:  file_get_contents(/usr/local/etc/ncp-config.d/.letsencrypt.sh.cfg): failed to open stream: No such file or directory in /var/www/ncp-web/elements.php on line 137
[Wed Mar 06 23:39:19.287192 2019] [php7:warn] [pid 3810] [client 192.168.1.3:59797] PHP Warning:  Invalid argument supplied for foreach() in /var/www/ncp-web/elements.php on line 21
[Wed Mar 06 23:39:19.290828 2019] [php7:notice] [pid 3810] [client 192.168.1.3:59797] PHP Notice:  Undefined index: app in /var/www/ncp-web/index.php on line 219
grep: /etc/mysql/mariadb.conf.d/90-ncp.cnf: No such file or directory
grep: /etc/mysql/mariadb.conf.d/90-ncp.cnf: No such file or directory
[Wed Mar 06 23:39:27.479334 2019] [php7:warn] [pid 3820] [client 192.168.1.3:59802] PHP Warning:  file_get_contents(/usr/local/etc/ncp-config.d/.letsencrypt.sh.cfg): failed to open stream: No such file or directory in /var/www/ncp-web/elements.php on line 183
find: '/etc/letsencrypt/live/': Permission denied

```
</details>

<details>
<summary>Database logs</summary>

```

2019-03-06 22:41:52 1970905920 [Note] /usr/sbin/mysqld: Shutdown complete

2019-03-06 22:41:54 1988828976 [Note] InnoDB: Using mutexes to ref count buffer pool pages
2019-03-06 22:41:54 1988828976 [Note] InnoDB: The InnoDB memory heap is disabled
2019-03-06 22:41:54 1988828976 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
2019-03-06 22:41:54 1988828976 [Note] InnoDB: GCC builtin __atomic_thread_fence() is used for memory barrier
2019-03-06 22:41:54 1988828976 [Note] InnoDB: Compressed tables use zlib 1.2.8
2019-03-06 22:41:54 1988828976 [Note] InnoDB: Using Linux native AIO
2019-03-06 22:41:54 1988828976 [Note] InnoDB: Using generic crc32 instructions
2019-03-06 22:41:54 1988828976 [Note] InnoDB: Initializing buffer pool, size = 370.0M
2019-03-06 22:41:54 1988828976 [Note] InnoDB: Completed initialization of buffer pool
2019-03-06 22:41:54 1988828976 [Note] InnoDB: Highest supported file format is Barracuda.
2019-03-06 22:41:54 1988828976 [Note] InnoDB: 128 rollback segment(s) are active.
2019-03-06 22:41:54 1988828976 [Note] InnoDB: Waiting for purge to start
2019-03-06 22:41:54 1988828976 [Note] InnoDB:  Percona XtraDB (http://www.percona.com) 5.6.41-84.1 started; log sequence number 6410853910
2019-03-06 22:41:54 1111487296 [Note] InnoDB: Dumping buffer pool(s) not yet started
2019-03-06 22:41:54 1988828976 [Note] Plugin 'FEEDBACK' is disabled.
2019-03-06 22:41:54 1988828976 [Note] Server socket created on IP: '127.0.0.1'.
2019-03-06 22:41:54 1988828976 [Note] /usr/sbin/mysqld: ready for connections.
Version: '10.1.37-MariaDB-0+deb9u1'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  Raspbian 9.0

```
</details>

<details>
<summary>Nextcloud logs</summary>

```

{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:02+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Invalidating remembered session tokens older than 2019-02-19T23:30:02+00:00","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:02+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Finished OC\\Authentication\\Token\\DefaultTokenCleanupJob job with ID 15 in 0 seconds","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:02+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Run OC\\Preview\\BackgroundCleanupJob job with ID 2680","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Finished OC\\Preview\\BackgroundCleanupJob job with ID 2680 in 1 seconds","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Run OCA\\DAV\\BackgroundJob\\UpdateCalendarResourcesRoomsBackgroundJob job with ID 2683","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Finished OCA\\DAV\\BackgroundJob\\UpdateCalendarResourcesRoomsBackgroundJob job with ID 2683 in 0 seconds","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Run OCA\\Support\\BackgroundJobs\\CheckSubscription job with ID 2686","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Finished OCA\\Support\\BackgroundJobs\\CheckSubscription job with ID 2686 in 0 seconds","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Run OCA\\UpdateNotification\\ResetTokenBackgroundJob job with ID 2517","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Finished OCA\\UpdateNotification\\ResetTokenBackgroundJob job with ID 2517 in 0 seconds","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Run OC\\Authentication\\Token\\DefaultTokenCleanupJob job with ID 2522","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Invalidating session tokens older than 2019-03-05T23:30:03+00:00","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Invalidating remembered session tokens older than 2019-02-19T23:30:03+00:00","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Invalidating session tokens older than 2019-03-05T23:30:03+00:00","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Invalidating remembered session tokens older than 2019-02-19T23:30:03+00:00","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Finished OC\\Authentication\\Token\\DefaultTokenCleanupJob job with ID 2522 in 0 seconds","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Run OC\\Log\\Rotate job with ID 2523","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Finished OC\\Log\\Rotate job with ID 2523 in 0 seconds","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Run OCA\\Files\\BackgroundJob\\ScanFiles job with ID 5","userAgent":"--","version":"15.0.5.3"}
{"reqId":"uIg6ATsVFzbOJLv6pcJb","level":0,"time":"2019-03-06T23:30:03+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Finished OCA\\Files\\BackgroundJob\\ScanFiles job with ID 5 in 0 seconds","userAgent":"--","version":"15.0.5.3"}

```
</details>

**Comment by nachoparker** [comment]
seems like you missed some important update in the past. `ncc` is an alias to `sudo -u www-data php /var/www/nextcloud/occ`

**Comment by nero120** [comment]
Was ncc created by Nextcloud or NCP? I might just do a refresh of ncp and then restore the db from backup, might solve some of the other issues I'm having.

**Comment by nachoparker** [comment]
by NCP. If you have missed updates it is always a good idea

**Comment by TimTaler007** [comment]
> by NCP. If you have missed updates it is always a good idea

I have a similar problem to update an old version of NCP (1.5.1) to current version 1.10.0.
It will stop the update with message:
`./update.sh: line 202: ncc: command not found`
So I tried to add the alias ncc with command `sudo -u www-data php /var/www/nextcloud/occ` but this alias does not work for sudo. How can I solve this problem or do a refresh of ncp?

**Comment by nachoparker** [comment]
better to backup, and restore in the latest image

**Comment by nisses** [comment]
The other way is to put the following content into /usr/local/bin/ncc file:
```
#!/bin/bash
sudo -u www-data php /var/www/nextcloud/occ "$@"
```
remember to `chmod +x /usr/local/bin/ncc` 

But... since ncc was introduced in v0.56.19 which was long time ago 2018-06-06 and so many people have this problem (and other problems caused by skipped/failed/unfinished updates) there should be a way to check the state of NextCloudPi and NextCloud for errors and inconsistency. Relational DB and the files and directory structure of NextCloud and NextCloudPi can be inspected by a script. There is ncc integrity:check-app and ncc integrity:check-core and ncc status but I think it does not cover all area of possible problems.

**Comment by TimTaler007** [comment]
> The other way is to put the following content into `/usr/local/bin/ncc` file:
> 
> ```
> #!/bin/bash
> sudo -u www-data php /var/www/nextcloud/occ "$@"
> ```
> remember to `chmod +x /usr/local/bin/ncc`
> 
>
This solved my update-problem with version 1.11.2.
Thank you so much for your help and support! I had no hope to solve it without a restore and found no time to do it. Your tip was easier to realize and works for the moment and the current version.

**Comment by nachoparker** [comment]
@nisses  there's an issue / improvement request to improve the update process because people skip updates even though they shouldn't.

But there is nobody working on it right now
